[lldb] Address mask sbprocess apis and new mask invalid const

jasonmolenda opened the pull request:
[lldb] Add SBProcess methods for get/set/use address masks (#83095)

I'm reviving a patch from phabracator, https://reviews.llvm.org/D155905
which was approved but I wasn't thrilled with all the API I was adding
to SBProcess for all of the address mask types / memory regions. In this
update, I added enums to control type address mask type (code, data,
any) and address space specifiers (low, high, all) with defaulted
arguments for the most common case.  I originally landed this via
https://github.com/llvm/llvm-project/pull/83095 but it failed on CIs
outside of arm64 Darwin so I had to debug it on more environments
and update the patch.

This patch is also fixing a bug in the "addressable bits to address
mask" calculation I added in AddressableBits::SetProcessMasks. If lldb
were told that 64 bits are valid for addressing, this method would
overflow the calculation and set an invalid mask. Added tests to check
this specific bug while I was adding these APIs.

This patch changes the value of "no mask set" from 0 to
LLDB_INVALID_ADDRESS_MASK, which is UINT64_MAX. A mask of all 1's
means "no bits are used for addressing" which is an impossible mask,
whereas a mask of 0 means "all bits are used for addressing" which
is possible.

I added a base class implementation of ABI::FixCodeAddress and
ABI::FixDataAddress that will apply the Process mask values if they
are set to a value other than LLDB_INVALID_ADDRESS_MASK.

I updated all the callers/users of the Mask methods which were
handling a value of 0 to mean invalid mask to use
LLDB_INVALID_ADDRESS_MASK.

I added code to the ABISysV_arm64 Fix override methods to apply the
Highmem masks if they have been set.  These will not be set on a
Linux environment, but I have tests in TestAddressMasks.py which
set the highmem masks and I need all Fix*Mask implementations to
check & use them for the tests to pass.

rdar://123530562

jasonmolenda committed:
[lldb] Add SBProcess methods for get/set/use address masks (#83095)

I'm reviving a patch from phabracator, https://reviews.llvm.org/D155905
which was approved but I wasn't thrilled with all the API I was adding
to SBProcess for all of the address mask types / memory regions. In this
update, I added enums to control type address mask type (code, data,
any) and address space specifiers (low, high, all) with defaulted
arguments for the most common case.

This patch is also fixing a bug in the "addressable bits to address
mask" calculation I added in AddressableBits::SetProcessMasks. If lldb
were told that 64 bits are valid for addressing, this method would
overflow the calculation and set an invalid mask. Added tests to check
this specific bug while I was adding these APIs.

rdar://123530562

jasonmolenda committed:
[lldb] Use new value for no-mask, add ABI base method, update others

This changes the value of "no mask set" from 0 to
LLDB_INVALID_ADDRESS_MASK, which is UINT64_MAX. A mask of all 1's
means "no bits are used for addressing" which is an impossible mask,
whereas a mask of 0 means "all bits are used for addressing" which
is possible.

I added a base class implementation of ABI::FixCodeAddress and
ABI::FixDataAddress that will apply the Process mask values if
they are set to a value other than LLDB_INVALID_ADDRESS_MASK.

I updated all the callers/users of the Mask methods which were
handling a value of 0 to mean invalid mask to use
LLDB_INVALID_ADDRESS_MASK.

I added code to the ABISysV_arm64 Fix override methods to
apply the Highmem masks if they have been set.  These will
not be set on a Linux environment, but I have tests in
TestAddressMasks.py which set the highmem masks and I need
all Fix*Mask implementations to check & use them for the
tests to pass.  It may be better to conditionalize the
highmem tests for arm64-apple, where that's an environment
we do need to support.

jasonmolenda commented:
@DavidSpickett I spent a little time debugging this on aarch64-ubuntu and also testing on x86_64-macos (which has no address masks at all).  This PR has the original commit from https://github.com/llvm/llvm-project/pull/83095 and then a second commit with the changes I needed to make as I debugged this in the full environment, so you only need to look at the second commit for the new changes.

The main question I have (beyond feedback on how I describe things in comments of course ;) ) is about how TestAddressMasks.py assumes all Fix*Address implementations will handle a low and high memory address mask.  The test currently assumes they do, but that means all people writing these methods need to get the highmem mask if it is a high-mem addr_t and apply that mask if it is set.  I think it's the right choice, but I'm not wedded to it; maybe the right choice is narrowing that part of TestAddressMasks.py to only run on darwin systems.

I also added a base class ABI::Fix*Address so these tests can be used on targets with no address masks at all, if someone wants to manipulate them via SB API for some reason (of course, the real reason is to make the test runnable).  Maybe the test should be limited to only running on aarch64-linux and arm64-apple, I also am not convinced my choice is best here.

DavidSpickett commented:
>  is about how TestAddressMasks.py assumes all Fix*Address implementations will handle a low and high memory address mask. The test currently assumes they do

I'm not sure it makes much difference unless handling low/high produces a large amount of boilerplate in every plugin.

It is a bit of a foot gun to be able to tell the plugin for, for example, Arm 32, that it has high/low mem masks. Since they won't make any sense, but it's pretty well hidden.

If somehow Arm/ppc/mips/intel/s390x grew a need to handle high/low mem and needed zero changes other than the mask, folks would be able to just set it without changing lldb (though I expect that wouldn't be the only change needed anyway).

jasonmolenda commented:
> > is about how TestAddressMasks.py assumes all Fix*Address implementations will handle a low and high memory address mask. The test currently assumes they do
> 
> I'm not sure it makes much difference unless handling low/high produces a large amount of boilerplate in every plugin.
> 
> It is a bit of a foot gun to be able to tell the plugin for, for example, Arm 32, that it has high/low mem masks. Since they won't make any sense, but it's pretty well hidden.

I agree, the more I think about this the less I'm delighted by making every plugin support high/low address masks when it may not make any sense there.  I like keeping the support in the SysV AArch64 ABI even though it's not used on Linux in case this ABI were used for generic AArch64 firmware debug scenario where someone might try to use that ABI plugin.  

In the API test, I can have a dedicated test for the highmem checks an only run it on aarch64/arm64 targets.  I can add a test for the non-aarch64/arm64 targets which tests that setting the highmem mask has no effect (and remove the highmem support from the base ABI method impl).

jasonmolenda committed:
Make baseclass impl only use a single address mask

I think the best behavior is that the highmem masks are no-ops for
all targets that don't have the concept of differet addressing masks
for high and low memory.  I removed the code from the base class
methods so they do not do this.  I updated all AArch64 ABI plugins
that provide Fix methods to use the highmem masks for high addresses
if there is one.

In the TestAddressMasks.py test, I moved all tests that would differ
depending on the target into their own test methods, with the majority
of the mask tests which aren't testing highmem issues run by all.

jasonmolenda commented:
I pushed an update to this PR where I have the base class Fix methods in ABI.cpp ignore the highmem masks completely, I added highmem mask use when it's a highmem address and they are set to the AArch64ABI class so all the AArch64 ABI plugins are checking this.  I changed the API test to only do highmem tests on AArch64 targets, and on other targets they will test that highmem masks are ignored.  If any target wants this feature, I would like them to have to explicitly declare it in the TestAddressMasks.py so things don't kinda-work/fail unintentionally.

jasonmolenda committed:
Have the AArch64 ABI Fix methods no-op if mask is unset

DavidSpickett commented on lldb/test/API/python_api/process/address-masks/TestAddressMasks.py:
```diff
@@ -0,0 +1,130 @@
+        )
+        mask = process.GetAddressMask(lldb.eAddressMaskTypeAny)
+        process.SetAddressMask(lldb.eAddressMaskTypeCode, mask | 0x3)
+        process.SetAddressMask(lldb.eAddressMaskTypeCode, 0xFFFFFC0000000003)
```
These two lines do the same thing?

jasonmolenda replied:
Ah, thanks for spotting that.  I originally wrote it as a hex literal and then I thought "Oh it would be clearer to bitmask in the 0b111" and forgot to remove the lit.

jasonmolenda committed:
Remove redudant mask setting, clear settings values

DavidSpickett reviewed (approved):
Still not sure this is the ideal  API but we can't know that until someone (aka you :) ) has used it for a bit, so let's get this in.

jasonmolenda commented:
> Still not sure this is the ideal API but we can't know that until someone (aka you :) ) has used it for a bit, so let's get this in.

Thanks for all the help in finishing this one.  I think this approach with enums to select the different types/ranges, with defaulted arguments so the most frequent use case is easy, is not bad.  My original patch had a lot of different methods for the different operations and that was a lot less clear for script writers I think.

jasonmolenda commented:
nb: this got a bot failure on the arm-linux-ubuntu bot,

```
            mask = process.GetAddressMask(lldb.eAddressMaskTypeAny)
            process.SetAddressMask(lldb.eAddressMaskTypeCode, mask | 0x3)
            self.assertEqual(
                0x000002950001F694,
                process.FixAddress(0x00265E950001F697, lldb.eAddressMaskTypeCode),
            )
    
    The API returned 0x000002950001f694 instead of the expected
    0x00265e950001f696.  The low bits differ because ABISysV_arm hardcodes
    the Code address mask to clear the 0th bit, it doesn't use the
    Process code mask.  I didn't debug why some of the high bytes were
    dropped.  
```

I'm skipping TestAddressMasks for arch=arm now, but I don't even know if we should be able to set a mask which preserves the high 32-bits of an addr_t on a 32-bit host; all masks have to mask off the high word of addr_t to be a valid addresss, at a minimum.  I don't have access to a 32-bit host that lldb runs on myself so my position is that I'll skip this API test entirely on 32-bit targets, I don't think the concepts make much sense there.

DavidSpickett commented:
We could have ABI plugins limit the mask to the known maximum address size, or set all the upper bits to 1. Though I think `0...0<lower 32 of mask>` is less confusing. But it's only fixing what you might call a visual issue, the mask will have upper 32 bits preserved, but at runtime no address will have anything there to preserve, so the result is still correct. We've been storing 32 bit addresses in a 64 bit addr_t for the lifetime of lldb, so if we had problems in this area we'd know already.

I'm inclined to keep it as is (it's simpler) and if people find the current behaviour confusing we can always make some of the plugins fix up the mask just so it looks correct.

(and also, what is correct, top 32 bits of 1s because you can't store them in a 32 bit address, or 0s because there are no bits to preserve anyway :shrug: )
